recipe/build-mpi.sh

```diff
@@ -31,7 +39,9 @@ export LIBRARY_PATH="$PREFIX/lib"
             --with-wrapper-cxxflags="-I$PREFIX/include" \
             --with-wrapper-fcflags="-I$PREFIX/include" \
             --with-wrapper-ldflags="-L$PREFIX/lib -Wl,-rpath,$PREFIX/lib" \
-            --with-sge
+            --with-sge \
+            $build_with_cuda
 
 make -j"${CPU_COUNT:-1}"
 make install
+echo opal_warn_on_missing_libcuda = 0 >> $PREFIX/etc/openmpi-mca-params.conf
```

Comment:
Is this still needed if we are only building with CUDA in certain cases?

Reply:
@leofang Follow @jakirkham advice, mess with the config var ONLY if `$build_with_cuda` is not the empty string `if [ -z "$build_with_cuda" ] ...`

Reply:
Thanks for confirming Lisandro! I wasn't actually sure how OpenMPI used this.

For more context `cudatoolkit` automatically becomes a dependency when `{{ compiler("cuda") }}` is added. So would expect this isn't needed in that case either and can just be dropped.

Though you know more about OpenMPI than I do. So please feel free to correct me if needed 🙂

Reply:
> Thanks for confirming Lisandro! I wasn't actually sure how OpenMPI used this.
> 

Setting it unconditionally would make no harm, but just in case, better safe than sorry.

> For more context `cudatoolkit` automatically becomes a dependency when `{{ compiler("cuda") }}` is added. So would expect this isn't needed in that case either and can just be dropped.
> 

Yes, @leofang confirmed that Open MPI access the CUDA runtime via `dlopen()` at runtime, thus being optional, so the CUDA runtime **SHOULD NOT** become a dependency of the openmpi package. However, I have no idea how to force `conda` to remove that implicit dependency. Any tips?

Reply:
Just for my own understanding, what is the motivation for not having a `cudatoolkit` dependency?

Reply:
Yes, I found that too. Seems like `conda install openmpi` would also install `cudatoolkit`...Would be preferable not to.

Reply:
> Just for my own understanding, what is the motivation for not having a `cudatoolkit` dependency?

Many MPI applications do not need GPU.

Reply:
Thanks for the context. When adding `{{ compiler("cuda") }}`, you will get 4 builds of `openmpi`. 3 with different CUDA versions and 1 that does not use CUDA. The one thing this recipe doesn't have right now is a way to select between the CUDA and non-CUDA installs. I'll make a suggestion in the diff as to how we can add that.

Reply:
@jakirkham maybe we can do this?
```
ignore_run_exports:
  - cudatoolkit
```

Reply:
Please see [this comment]( https://github.com/conda-forge/openmpi-feedstock/pull/54#issuecomment-577376632 ). For a suggested way to address this.

Reply:
Sorry I wrote that comment before I saw yours, Leo.

Unfortunately `ignore_run_exports` isn't likely to do what we want here. In the 1 CPU build `cudatoolkit` is already ignored. In the 3 CUDA builds this will result in packages that have linked to some version of CUDA, but no longer depend on `cudatoolkit`. This means users would need to install this `cudatoolkit`. Further they wouldn't know what version of CUDA packages were built with and would lack a way to select packages that mean their CUDA requirements. This would complicate deployment to a system that expects a specific version of CUDA (due to driver versions or other constraints).

Reply:
> Sorry I wrote that comment before I saw yours, Leo.

No worries, John.  I realized that too. 😄 

> Unfortunately `ignore_run_exports` isn't likely to do what we want here. In the 1 CPU build `cudatoolkit` is already ignored. In the 3 CUDA builds this will result in packages that have linked to some version of CUDA, but no longer depend on `cudatoolkit`. This means users would need to install this `cudatoolkit`. 

1. The linking is done in runtime, so technically it might be even possible that `openmpi` built with CUDA 10.2 could work with CUDA 9.2 discovered in runtime. (@jsquyres I am crying out again for confirmation 😭)
2. We do expect users to install `cudatoolkit` explicitly, either by themselves or by the constraint of some other packages like `cupy` or `pytorch`.

> Further they wouldn't know what version of CUDA packages were built with and would lack a way to select packages that mean their CUDA requirements. This would complicate deployment to a system that expects a specific version of CUDA (due to driver versions or other constraints).

If item 1 above is true, this would not be any problem. If not, then we do need a selector or your proposal below to enforce this requirement.

Reply:
Ok thanks for that additional context. Let's wait and see what others from the OpenMPI world say then 🙂

Reply:
Oops, sorry -- missed this prompting for information -- thanks to @jakirkham for poking me with a pointer to this comment.

> Yes, @leofang confirmed that Open MPI access the CUDA runtime via dlopen() at runtime, thus being optional, so the CUDA runtime SHOULD NOT become a dependency of the openmpi package. 

That is correct.  I haven't read this entire lengthy conversation, but I think that our reasons for doing that were for almost exactly the use case of this PR: we want CUDA to be able to be used, but some clusters only have funding for GPUs on some of their nodes (and therefore don't have the CUDA libraries/etc. installed on all nodes in their cluster).  But they still want a single Open MPI installation to be able to be used on _all_ nodes -- both with and without the CUDA libraries/etc.  So we made some `cuda` components in Open MPI, but they manually try to `dlopen("libcuda.so.1", ...)` at run time, as opposed to having an `ld`-mandated requirement for finding that library at run time.

That's what the `opal_warn_on_missing_cuda` MCA param is for: if you configured Open MPI to expect CUDA support, we warn if we can't find it at run time.  BUT: if you have a scenario like I outlined above (e.g., one Open MPI install that spans nodes with and without CUDA), then it can be ok if Open MPI doesn't find the CUDA library at run time (e.g., if it's running on a node that doesn't have a GPU / doesn't have CUDA).  So you found exactly the right MCA param to shut Open MPI up about this warning.

Does this answer your question?

Reply:
Also:

> The linking is done in runtime, so technically it might be even possible that openmpi built with CUDA 10.2 could work with CUDA 9.2 discovered in runtime. (@jsquyres I am crying out again for confirmation 😭)

Open MPI is specifically looking for `libcuda.so.1` or `libcuda.dylib`.  So even if you have CUDA 9.2 or CUDA 10.2, if they have `libcuda.so.1`, it should be ok.  I.e., I _assume_ that NVIDIA provides the usual ABI guarantees about SO version numbers.